### ISSUE TITLE
Using metadata causes bad line number node placement

### ISSUE THREAD

**Issue by maleadt** [JuliaLang/julia#14346]
When I combine certain functions with a `:meta` expression (eg. using `@noinline`), the line number node detection during `emit_function` fails:

```julia
@noinline function foobar(id, mask, limit)
    if id <= limit && mask[id]
        return true
    end
end

foobar(1, [true], 1)

functionloc(foobar)
```
```
ERROR: LoadError: could not determine location of method definition
```

(even though `functionloc` seems less sophisticated than the actual detection in `codegen.cpp`, in this case the behaviour is the same and `filename` is assigned `no file`)

I'm not familiar with the lowering logic of `inference.jl`, but the cause seems to be a `newvar` placed at the top of the function: when not using `:meta` attribute, the first line number node is placed before the `newvar`, but when prepending a `:meta` the line node sticks below the `newvar`, so `skip_meta` doesn't work.

Example for this function, using `@noinline`:

```
Expr(:body,
  <newvar :#s1>,
  Expr(:meta, :noinline)::Any,
  # line 2 /tmp/test.jl,
  ...
```

Without `@noinline`:
```
Expr(:body,
  # line 2 /tmp/test.jl,
  <newvar :#s1>,
  ...
```

**Comment by maleadt** [comment]
Turns out this specific case was fixed by 464904e17256e0e3fdb9f50d0d401abf3d8547d1 (specifically, 40f3859d065c71f79cbca3f6100bfa1ea033d7ac), after which no `NewVarNode` is put before the `:meta` Expr anymore. Not sure whether that was a systemic fix; should I add a test based on this previously failing snippet?
cc @JeffBezanson

**Comment by tkelman** [comment]
unless you wouldn't mind it potentially regressing, I think a test would be great
